### DISCUSSION TITLE
Please help a noob with Machine Learning challenge

### DISCUSSION THREAD

**Discussion by akshayrd1** [community/community#82747] [Programming Help]
Please help a noob with machine learning challenge. I am manufacturing technology engineer working at metal manufacturing plant.
I have data set of actual chemical composition (comprising 8 elements) of many castings of a steel grade.
I also have measured volume percent of a microstrutural phase for each casting of this steel grade.
I believe there is a relationship between chemical composition and volume percent of the phase. 
I want to determine this relationship and formulize it.
That way, based on desired volume percent of microstructural phase, I can decide the chemical composition of casting.
I think machine learning can help me with this. Please guide me.

**Comment by Gouravlohar** [comment]
I understand that and Machine Learning will definitely help to find relationship between Chemical Composition and Volume percent of the phase. You can follow these steps :

**1.  Feature Engineering
2.  Find Missing Value (If You find any)
3.  Split dataset into training, validation and testing sets**

as per Your question you want continuous value so Regression Model will help you better.

**4. You can try any of these Regression Models:** 
i) Linear Regression (For Simple Datasets)
ii) Gradient Boosting regression
iii) Random Forest Regressor 
5. Train and evaluate your model 
6.  After that You can be able to find Your relationship
7.  Once satisfied with your model performance, use it to predict the chemical composition.
 Thats it

**Note: - Take care of Overfitting and ensure your model generalizes well on unseen data **

**_Example Code_** 
``import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error

#If you have data in csv file use pandas library to import them
df=pd.read_csv('your_file_path')

# Assuming X is your feature matrix (chemical composition) and y is your target variable (volume percent)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Choose a model
model = RandomForestRegressor()

# Train the model
model.fit(X_train, y_train)

# Make predictions
predictions = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, predictions)
print(f'Mean Squared Error: {mse}')

# Feature importance analysis
feature_importances = model.feature_importances_
print('Feature Importances:', feature_importances)
` `

Good Luck!

**Comment by queenofcorgis** [comment]
Thanks for posting in the GitHub Community, @akshayrd1 ! 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.

**Comment by zippynx** [comment]
Hello @akshayrd1! Welcome to the ML world. As a manufacturing engineer, you've hit on a classic optimization problem.

The previous answer is good for predicting the Volume % based on the Composition (Forward problem). However, you mentioned wanting to decide the chemical composition based on a desired volume percent (Inverse problem). Predicting 8 variables directly from 1 variable is mathematically tricky because multiple different compositions might yield the exact same volume percent.

Here is the standard Machine Learning approach for your specific manufacturing use case:

Step 1: Train a Forward Regression Model
First, train a model where X = 8 chemical elements and y = Volume %.
Algorithms like Support Vector Regression (SVR) are excellent for this because they handle complex, non-linear relationships in metallurgical data very well and are robust against outliers.

Step 2: Optimize for the Desired Composition
Once your model accurately predicts the Volume % from the chemistry, you use an optimization algorithm to find the exact 8 elements that will output your target Volume %.

Here is a simplified Python code example to demonstrate this pipeline:
import pandas as pd
import numpy as np
from sklearn.svm import SVR
from sklearn.preprocessing import StandardScaler
from scipy.optimize import minimize

# 1. Load and Prep Data
# df = pd.read_csv('your_casting_data.csv')
# X = df[['elem1', 'elem2', 'elem3', 'elem4', 'elem5', 'elem6', 'elem7', 'elem8']]
# y = df['volume_percent']

# Scale data (Crucial for distance-based models like SVR)
scaler_X = StandardScaler()
X_scaled = scaler_X.fit_transform(X)

# 2. Train the Model 
# You can tune parameters like C, gamma, and epsilon for better accuracy
model = SVR(kernel='rbf', C=100, gamma='scale', epsilon=0.1)
model.fit(X_scaled, y)

# 3. Optimization: Find composition for a TARGET volume %
target_volume = 45.0 # Example: You want exactly 45% volume phase

def objective_function(composition):
    # We want to minimize the difference between the model's prediction and your target
    comp_scaled = scaler_X.transform([composition])
    predicted_vol = model.predict(comp_scaled)[0]
    return abs(predicted_vol - target_volume)

# Initial guess (e.g., the average chemical composition from your historical data)
initial_guess = X.mean().values

# Define bounds based on your plant's realistic chemical limits for each element
# Example for 8 elements: [(min, max), (min, max), ...]
bnds = [(0.1, 2.0), (0.05, 0.5), (0.0, 1.0), (0.0, 1.0), (0.0, 1.0), (0.0, 1.0), (0.0, 1.0), (0.0, 1.0)] 

result = minimize(objective_function, initial_guess, bounds=bnds, method='SLSQP')

print(f"Optimal Chemical Composition to achieve {target_volume}% is:")
print(result.x)

Why this works better for you: It mimics your real-world plant process. You input your desired target, and the algorithm finds the optimal chemical recipe within your specific manufacturing limits!

Let me know if you need help tuning the SVR parameters or setting up the boundaries for your specific steel grade. Happy to help!

**Comment by blaze0089** [comment]
---

Hey, so both answers above are really helpful for the prediction side of things, but I noticed something — you said you want to **formulize** it. Like an actual equation you can write down and hand to your team. That's a completely different goal and neither answer really covers it.

What you're looking for is called **Symbolic Regression**. Instead of giving you a black box that just spits out numbers, it actually hunts for a real mathematical formula that fits your data. Something like:

`Volume% = 1.4 × Carbon² + 0.6 × Manganese − 0.3 × Chromium`

You can put that straight into Excel, share it with colleagues, use it on the shop floor — no Python needed at all. For a manufacturing engineer that's way more practical than a model nobody else can use.

The library that does this best right now is called PySR, here's how simple it is to get started:

```python
pip install pysr
```

```python
from pysr import PySRRegressor

model = PySRRegressor(
    niterations=100,
    binary_operators=["+", "-", "*", "/"],
    unary_operators=["square", "sqrt"],
)

model.fit(X, y)
print(model)
```

That last `print(model)` literally shows you the formula it found. You can then look at it and say "yeah that makes metallurgical sense" or "that's wrong" based on your own domain knowledge — which is something a Random Forest can never give you.

One more thing since manufacturing datasets tend to be on the smaller side — instead of a basic train/test split, use k-fold cross validation so every single casting in your dataset pulls its weight:

```python
from sklearn.model_selection import cross_val_score
scores = cross_val_score(model, X, y, cv=5, scoring='r2')
print(f"Average R2: {scores.mean():.3f}")
```

The optimization approach in the answer above is still the right move for the inverse problem (going from target volume% back to composition), so combine both — use symbolic regression to get your formula first, then use that formula inside the optimizer instead of a black box SVR. That way everything is transparent and explainable.

Feel free to share how many castings you have in your dataset, that'll help figure out the best approach for your specific situation!

---

**Comment by josephwatson5237-web** [comment]
This is essentially a multivariate regression problem, start with feature exploration and try models like linear regression, random forest, or gradient boosting to map chemical composition to phase volume. It’s similar in structure to decision modeling seen in [property development capital raising services](https://sl1nk.com/1bn2ptz), where multiple inputs are used to predict constrained outcomes.
